### ISSUE TITLE
task1_ai_ml_engg : NLU algorithm to be used by diagnosis/assessment bot 

### ISSUE THREAD

**Issue by nlauchande** [omdena/ptsd#4]

**Comment by MamadouBousso** [comment]
**Part 1: Intent recognition using RASA NLU
Intent recognition:  How to better understand your users and what they say ?** 

**1. RASA NLU** 
It uses a pipeline to process a user message sequentially and ultimately leads to the classification of user messages into intents and extraction of entities. The pipeline is defined by the following stages: tokenization, featurization, intent classification, entity extraction, pattern matching. The following image is an example from https://blog.rasa.com/enhancing-rasa-nlu-with-custom-components/:
<img width="1199" alt="Capture d’écran 2019-07-09 à 20 01 27" src="https://user-images.githubusercontent.com/1431116/60919361-06cda000-a285-11e9-8b3a-4eeb74c11b4c.png">
Once the pipeline is defined, each component is called one after another and produces an output which is either added to the RASA NLU model output or used as an input for another component. We can see the following schema from https://blog.rasa.com/enhancing-rasa-nlu-with-custom-components/:
<img width="704" alt="Capture d’écran 2019-07-09 à 07 54 20" src="https://user-images.githubusercontent.com/1431116/60919705-ee11ba00-a285-11e9-982f-efca5ebb4b39.png">

**- How The pipeline works ?**

A so-called "context" plays the role of information dispatcher between components and there are three main stages:

- Initialization of the components before training

- The components train itself based on the context and outputs of previous components

- Save the trained components for future use

- Creation of metadata of the model which describe the overall NLU model

**- Two types of components**

- Pre-built RASA components 

- Custom components

**2. Intent recognition**
RASA uses the concept of _intents_ in order to categorize a message. RASA NLU will classify user messages following one or multiple intents. Two major components are used for this task:

- **Pretrained Embeddings( intent_classifier_sklearn)**

It uses the **SPACY** library to load pretrained language model which are used to represent each word of the user message as a dense numeric vector. It captures the semantic and syntactic meaning of a word.
__PROS:_
_- benefit from research advances in training more powerful and meaningful word embeddings_
_- Require only little training to make confident intent predictions_
_- get robust classification results with a small amount of data_
_- fast training__
_CONS_
_- good word embeddings are not available for all languages_
_- don't cover domain-specific words_

- **Supervised-embeddings( Intent classifier Tensorflow embeddings)**
Learns embeddings for all the intents and the words simultaneously. The embeddings are used to rank the similarity between an input sentence and all the intents.
Used with the **intent_featurizer_count_vectors** component which counts how often distinct word of your training data appear in a message and provide that as input for the intent classifier
_-PROS_
_-Adapts to your domain specific message_
_-Inherently language independant_
_-support messages with multiple intents_
_-CONS_
_- Need more training data_

**3. Best fit for our problem**

- Pre-trained word embeddings exist for the English language

- We need support for multiple intents per message

- We have domain specific terms

Based on the remarks above:

- if we have a lot of training data we can use Supervised Embeddings
- Otherwise, we can start with Pretrained Embeddings

**4. Research papers**
[Word2vecpaper.pdf](https://github.com/omdena/ptsd/files/3374858/Word2vecpaper.pdf)
[starspace.pdf](https://github.com/omdena/ptsd/files/3374862/starspace.pdf)
https://rasa.com/docs/rasa/nlu/components/#embeddingintentclassifier
https://spacy.io/
https://rasa.com/docs/rasa/nlu/components/#countvectorsfeaturizer

**Comment by LaurenSpiegel** [comment]
I am working on swapping in fastai text for the intent classifier in rasa so I can test its performance.

Helpful links:
- Blog on enhancing rasa nlu with custom components: https://blog.rasa.com/enhancing-rasa-nlu-with-custom-components/
- Blog on training and testing nlu part of rasa in jupyter notebook: https://towardsdatascience.com/building-a-conversational-chatbot-for-slack-using-rasa-and-python-part-1-bca5cc75d32f

Last night I trained a language model on a dataset of humans texts to bots--
https://github.com/LaurenSpiegel/rasa-fastai/blob/master/notebooks/train-language-model.ipynb

Next step is to figure out how to incorporate this into rasa NLU using their Component class to train and run as an intent classifier.

**Comment by LaurenSpiegel** [comment]
Update:
1) I retrained a language model with additional examples to be used by a classifer and then also trained a classifer -- https://github.com/LaurenSpiegel/rasa-fastai/blob/master/notebooks/language-model-and-classifier.ipynb
2) Started the rasa integration -- https://github.com/LaurenSpiegel/rasa-fastai/blob/master/notebooks/integrate-into-rasa.ipynb

Unfortunately hit some python package integration issues so didn't get as far as I would have liked :(

**Comment by rrupam** [comment]
Hi @LaurenSpiegel Thanks for sharing your notebook. It's a great place to begin. I went through your code and I had two follow up questions:
1. From reading the FastAI ULMFiT description, it seems like we can use their pretrained model and supply it with limited data to fine tune for our own classification task. Am I right in understanding that you're doing this in your code, too? If yes, can we use the dataset provided by Colton, label it manually and feed it in to get our classification model? And how limited can the data be.. are we talking maybe a 100 posts with labels or is it more? 
2. My second question is regarding the paper discussed in task #7 : This paper also deals with classification of posts into one of 15 classes, based on CBT. They have their own model (CNN+GLoVe), but if we could get the data that they used, we could feed it into your model and see how it works?

**Comment by LaurenSpiegel** [comment]
@rrupam, fastai ulmfit uses transfer learning. I am fine-tuning the language model using a chat dataset I found and then fine-tuning the classifier on just the rasa intents from their demo bot. You could fine-tune the language model and the classifier with whatever dataset you have. For the classifier, I have seen it work with a small dataset of about 100 examples so long as the number of classes is small and the dataset is balanced. I'm unclear on how you would use the dataset provided by Colton. What would the classes be and how would it be broken up?

**Comment by rrupam** [comment]
HI @LaurenSpiegel - Thank you for your explanation.

Essentially, I'm trying to figure out how best to classify the 'thinking errors' given a user's post and negative take. For eg.
Post: I really miss my ex! She just stopped talking to me all of the sudden, I don't know what happened
Negative Take: I won't get to talk to her again
Thinking Errors: Jumping to negative conclusions, Low Frustration Tolerance

This classification is based on CBT, which has 15 thinking error classes. I'm inclined to use this, as the example use case in the problem statement of the slack follows a conversation in which the therapist tries to identify the thinking error of the patient and accordingly guides the conversation. Thus, using NLU for classification into CBT thinking errors seems like a reasonable next step. In Issue #7, we discussed a paper that does exactly this. However, I think their model CNN+GLoVE may not be the best model. I think the language model you use can give better results. 

Regarding Colton's dataset -- I was first thinking that we could manually pick out a few dialogues from each conversation and label them with thinking errors. But I had a second look at the data ... and I agree with you--  I think this might not be the best way to do it. Picking the right dialogue itself is too time consuming and labeling it with the right thinking error is an expert's job, which we have in short supply. 

Ok, so going back to the paper in Issue #7, they have a corpus of ~500k posts + negative takes, a subset of which have been labeled by expert psychologists. We could get that data and feed it into your model, right? 
If this doesn't work -- any other labeled data set with posts (+negative take)+ thinking error label should work, right? How much labeled data do you think we will need to train this, considering that there are 15 classes? 
I'm interested to know what you think of this approach. :)

**Comment by LaurenSpiegel** [comment]
@rrupam , seems it should work for the 15 thinking error classes. I don't know how many examples you would need to get a good result without overfitting but might as well try with the sample you have and see how it works. You should be able to basically run my notebook with a few changes for the data.

**Comment by LaurenSpiegel** [comment]
Update:
1) I was able to integrate my fastai model into rasa and test --
https://github.com/LaurenSpiegel/rasa-fastai/blob/master/notebooks/integrate-into-rasa2.ipynb
2) I didn't properly split the data for training and testing since I wanted to just get it working. I will have to go back and train with a subset of data and then run the tests (I will also have to better handle the way the data is formatted for named entity recognition).

**Comment by nlauchande** [comment]
@LaurenSpiegel  : This looks awesome can you set up a demo for the team at your earliest convenience.
